### ISSUE TITLE
Database Schema Grammars Macroable

### ISSUE THREAD

**Issue by SebastianS90** [laravel/ideas#1162]
I would like to add custom column types in my migrations.

The `Blueprint` class is already Macroable, so it is easy to add a new type there. Now I need is to add the `typeXyz()` functions to all grammars I would like to support. Extending the grammar is not possible without also extending the corresponding connection. So it would be nice if the grammars in `Illuminate\Database\Schema\Grammars` could be macroable, too. Probably each grammar individually needs to pull in the Macroable trait, such that I can give a different implementation for each database system.

This is a better solution than raw statements in migrations, because I can support different database systems.

**Comment by ragingdave** [comment]
Would you be able to give of an example of what you are looking to accomplish? I would say that perhaps you are trying to add functionality that shouldn't exist in the Grammars. I would take Grammars to be something that is specific to the connection/database you are using. I guess an example would be adding a complex type that has different underlying column types? but even then.... you could end up making your type incompatible with all database connections.

**Comment by SebastianS90** [comment]
Example:

```php
// This already works, but results in an exception because the grammar has no `typeBinary16` method:
Illuminate\Database\Schema\Blueprint::macro('binary16', function ($column) {
    $this->addColumn('binary16', $column);
});

// This would be nice:
Illuminate\Database\Schema\Grammars\MySqlGrammar::macro('typeBinary16', function() {
    return 'binary(16)';
});
```

In my case, I like the mysql `binary(16)` type because it is the optimal type to store an [inet_pton](http://php.net/manual/en/function.inet-pton.php) encoded ip address. Laravel's `binary` migration type creates a `blob` instead, which is not as performant as a fixed size binary in mysql.

Making `MySqlGrammar` (and also the other schema grammars) macroable provides an easy way to add custom types.

**Comment by ragingdave** [comment]
I would almost argue that it might be more prudent to actually make the type binary.....be binary and not a blob. I'm not sure what the original reason was for using blob instead of binary and add an additional blob type.

Since this is potentially the only use case, because it's the only type I can see that is straight up changed from the default mysql types, I would probably argue against being able to macro db types. It just really doesn't feel like a place that should be macroable, for the reason of you can already do 99% of what's potentially needed (custom Blueprint macros). Additionally, It introduces potential fragmentation into the package space where if a package has a custom type, then it needs to add that type to any databases it would support rather than supporting all db's out of the box. Honestly I would just do something similar to https://github.com/spatie/laravel-binary-uuid and potentially try and start the conversation around why binary is blob and not binary in the first place.

**Comment by SebastianS90** [comment]
I strongly recommend against changing existing types, because it breaks a lot. Databases that initially were migrated using an old version of laravel have different column types than freshly migrated databases. This should never happen and is the reason why you should not edit existing migrations (rather add new migrations that change your columns as required).

Threre are probably more types or other DB specific features, e.g. in postgres. The Laravel core should only support types that can be used in all supported database systems. A custom application however should be able to add custom types and thereby drop support for other database systems.

The uuid package uses the typical child class approach. This is ok, but has two downsides:
- it is more code than just some `::macro`
- you cannot combine two such packages that both want to replace the grammar with their own subclass

**Comment by SebastianS90** [comment]
Implemented in laravel/framework#24513 (Schema Grammars extend the Database Grammar) :tada:
